File: utils/onebot/messages.py
"""OneBot message parsing without plugin-specific prompts or policies."""

from nonebot.adapters.onebot.v11 import Message, MessageSegment
from nonebot.adapters.onebot.v11.helpers import extract_image_urls


def normalize_message_segments(content):
    if isinstance(content, str):
        return [{"type": "text", "data": {"text": content}}] if content else []
    if isinstance(content, (MessageSegment, dict)):
        content = [content]
    if not isinstance(content, (list, tuple, Message)):
        return []
    result = []
    for item in content or []:
        if isinstance(item, MessageSegment) or (
            isinstance(getattr(item, "type", None), str)
            and isinstance(getattr(item, "data", None), dict)
        ):
            result.append({"type": item.type, "data": dict(item.data)})
        elif isinstance(item, dict) and "type" in item:
            result.append({"type": item["type"], "data": dict(item.get("data") or {})})
        elif isinstance(item, (list, tuple, Message, str)):
            result.extend(normalize_message_segments(item))
    return result


def mentioned_users(message):
    return [
        int(s["data"]["qq"])
        for s in normalize_message_segments(message)
        if s["type"] == "at" and str(s["data"].get("qq", "")).isdigit()
    ]


def find_forward_id(message):
    return next(
        (
            str(s["data"]["id"])
            for s in normalize_message_segments(message)
            if s["type"] == "forward" and s["data"].get("id")
        ),
        None,
    )


def image_urls(event, *, reply_first=False):
    reply = getattr(event, "reply", None)
    sources = [event.get_message(), reply.message if reply else Message()]
    if reply_first:
        sources.reverse()
    for source in sources:
        urls = extract_image_urls(source)
        if urls:
            return urls
    return []


def reply_image_segments(event):
    """Only images in the replied message; never silently switch to current input."""
    reply = getattr(event, "reply", None)
    return list(reply.message["image"]) if reply is not None else []

File: utils/onebot/messages.go
package onebot

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"time"

	"github.com/HakuchumuHYX/HakuBot/utils/network"
	zero "github.com/wdvxdr1123/ZeroBot"
	"github.com/wdvxdr1123/ZeroBot/message"
)

func MentionedUsers(msg message.Message) []int64 {
	var ids []int64
	for _, s := range msg {
		if s.Type == "at" {
			if id, err := strconv.ParseInt(s.Data["qq"], 10, 64); err == nil {
				ids = append(ids, id)
			}
		}
	}
	return ids
}
func ForwardID(msg message.Message) string {
	for _, s := range msg {
		if s.Type == "forward" {
			return s.Data["id"]
		}
	}
	return ""
}
func ImageURLs(msg message.Message) []string {
	var urls []string
	for _, s := range msg {
		if s.Type == "image" && s.Data["url"] != "" {
			urls = append(urls, s.Data["url"])
		}
	}
	return urls
}
func ReplyImages(ctx *zero.Ctx) message.Message {
	for _, s := range ctx.Event.Message {
		if s.Type == "reply" {
			msg := ctx.GetMessage(s.Data["id"])
			var result message.Message
			for _, part := range msg.Elements {
				if part.Type == "image" {
					result = append(result, part)
				}
			}
			return result
		}
	}
	return nil
}
func EventImageURLs(ctx *zero.Ctx, replyFirst bool) []string {
	if replyFirst {
		if urls := ImageURLs(ReplyImages(ctx)); len(urls) > 0 {
			return urls
		}
	}
	if urls := ImageURLs(ctx.Event.Message); len(urls) > 0 {
		return urls
	}
	if !replyFirst {
		return ImageURLs(ReplyImages(ctx))
	}
	return nil
}

type ForwardItem struct {
	Content message.Message
	Name    string
	UserID  int64
}
type Target struct{ GroupID, UserID int64 }
type ForwardStatus string

const (
	Sent           ForwardStatus = "sent"
	FallbackSent   ForwardStatus = "fallback_sent"
	TimeoutUnknown ForwardStatus = "timeout_unknown"
)

func SendForward(ctx context.Context, bot *zero.Ctx, target Target, items []ForwardItem, fallback bool) (ForwardStatus, error) {
	if (target.GroupID == 0) == (target.UserID == 0) || len(items) == 0 {
		return "", errors.New("one target and nonempty forward items are required")
	}
	ctx, cancel := context.WithTimeout(ctx, 60*time.Second)
	defer cancel()
	nodes := message.Message{}
	for _, item := range items {
		nodes = append(nodes, message.CustomNode(item.Name, item.UserID, item.Content))
	}
	action := "send_group_forward_msg"
	key := "group_id"
	id := target.GroupID
	if id == 0 {
		action = "send_private_forward_msg"
		key = "user_id"
		id = target.UserID
	}
	rsp := bot.CallActionWithContext(ctx, action, zero.Params{key: id, "messages": nodes})
	// ZeroBot reports transport failures as an empty response; never duplicate an unknown delivery.
	if rsp.Status == "" {
		return TimeoutUnknown, nil
	}
	if rsp.RetCode == 0 && rsp.Status == "ok" {
		return Sent, nil
	}
	if !fallback {
		return "", fmt.Errorf("forward rejected (%d): %s", rsp.RetCode, rsp.Message)
	}
	sent := 0
	for i, item := range items {
		action = "send_group_msg"
		if target.GroupID == 0 {
			action = "send_private_msg"
		}
		r := bot.CallActionWithContext(ctx, action, zero.Params{key: id, "message": item.Content})
		if r.Status != "ok" || r.RetCode != 0 {
			return "", fmt.Errorf("forward fallback stopped after %d messages: %s", sent, r.Message)
		}
		sent++
		if i < len(items)-1 {
			if err := network.Wait(ctx, time.Second); err != nil {
				return "", err
			}
		}
	}
	return FallbackSent, nil
}
